From f6ed9fa8a0ca5a7378eab0c8ba26faee314f7d0a Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?M=C3=A5rten=20Nordheim?= <marten.nordheim@qt.io>
Date: Mon, 6 Oct 2025 16:01:03 +0200
Subject: [PATCH 345/553] Http2: Relay an error for GOAWAY with NO_ERROR

It is intended to signal graceful shutdown.
Up to now we were unsetting any error for this GOAWAY reply.

Bug report says this was not seen before 6.9, so only picking to 6.10.

Conflicts resolved in cherry-pick to 6.10:
    - 74adf8c4a84d04f054251437255ca7d73977e23b was not picked back,
      which changed QScopedPointer to std::unique_ptr.

Fixes: QTBUG-139692
Change-Id: Ie46d8dca1f174f04bf55fe8087f4087566e9a2c9
Reviewed-by: Mate Barany <mate.barany@qt.io>
(cherry picked from commit 2192d2cfb9cd25d5907ceeb6fd3069e5a572af53)
Reviewed-by: Edward Welbourne <edward.welbourne@qt.io>
---
 src/network/access/http2/http2protocol.cpp    |  4 ++--
 tests/auto/network/access/http2/http2srv.cpp  |  5 +++--
 tests/auto/network/access/http2/http2srv.h    |  3 ++-
 tests/auto/network/access/http2/tst_http2.cpp | 14 ++++++++++----
 4 files changed, 17 insertions(+), 9 deletions(-)

diff --git a/src/network/access/http2/http2protocol.cpp b/src/network/access/http2/http2protocol.cpp
index c508cffef10..050beacb31c 100644
--- a/src/network/access/http2/http2protocol.cpp
+++ b/src/network/access/http2/http2protocol.cpp
@@ -104,8 +104,8 @@ void qt_error(quint32 errorCode, QNetworkReply::NetworkError &error,
 
     switch (http2Error) {
     case HTTP2_NO_ERROR:
-        error = QNetworkReply::NoError;
-        errorMessage.clear();
+        error = QNetworkReply::RemoteHostClosedError;
+        errorMessage = "Remote host signaled shutdown"_L1;
         break;
     case PROTOCOL_ERROR:
         error = QNetworkReply::ProtocolFailure;
diff --git a/tests/auto/network/access/http2/http2srv.cpp b/tests/auto/network/access/http2/http2srv.cpp
index 39b51a18c14..425d0a317da 100644
--- a/tests/auto/network/access/http2/http2srv.cpp
+++ b/tests/auto/network/access/http2/http2srv.cpp
@@ -133,11 +133,12 @@ void Http2Server::setSendTrailingHEADERS(bool enable)
     sendTrailingHEADERS = enable;
 }
 
-void Http2Server::emulateGOAWAY(int timeout)
+void Http2Server::emulateGOAWAY(int code, int timeout)
 {
     Q_ASSERT(timeout >= 0);
     testingGOAWAY = true;
     goawayTimeout = timeout;
+    goawayCode = code;
 }
 
 void Http2Server::redirectOpenStream(quint16 port)
@@ -414,7 +415,7 @@ void Http2Server::triggerGOAWAYEmulation()
     auto timer = new QTimer(this);
     timer->setSingleShot(true);
     connect(timer, &QTimer::timeout, [this]() {
-        sendGOAWAY(quint32(connectionStreamID), quint32(INTERNAL_ERROR), 0);
+        sendGOAWAY(quint32(connectionStreamID), quint32(goawayCode), 0);
     });
     timer->start(goawayTimeout);
 }
diff --git a/tests/auto/network/access/http2/http2srv.h b/tests/auto/network/access/http2/http2srv.h
index 3496233eb21..3b15278ce6f 100644
--- a/tests/auto/network/access/http2/http2srv.h
+++ b/tests/auto/network/access/http2/http2srv.h
@@ -75,7 +75,7 @@ public:
     void setRedirect(const QByteArray &redirectUrl, int count);
     // Send a trailing HEADERS frame with PRIORITY and END_STREAM flag
     void setSendTrailingHEADERS(bool enable);
-    void emulateGOAWAY(int timeout);
+    void emulateGOAWAY(int goawayCode, int timeout);
     void redirectOpenStream(quint16 targetPort);
 
     bool isClearText() const;
@@ -186,6 +186,7 @@ private:
 
     bool testingGOAWAY = false;
     int goawayTimeout = 0;
+    int goawayCode = 0;
 
     // Clear text HTTP/2, we have to deal with the protocol upgrade request
     // from the initial HTTP/1.1 request.
diff --git a/tests/auto/network/access/http2/tst_http2.cpp b/tests/auto/network/access/http2/tst_http2.cpp
index 385855e9130..45a9480ccff 100644
--- a/tests/auto/network/access/http2/tst_http2.cpp
+++ b/tests/auto/network/access/http2/tst_http2.cpp
@@ -604,8 +604,12 @@ void tst_Http2::goaway_data()
         QSKIP("This test requires TLS with ALPN to work");
 
     QTest::addColumn<int>("responseTimeoutMS");
-    QTest::newRow("ImmediateGOAWAY") << 0;
-    QTest::newRow("DelayedGOAWAY") << 1000;
+    QTest::addColumn<int>("goawayCode");
+    QTest::addColumn<QNetworkReply::NetworkError>("expectedError");
+    QTest::newRow("ImmediateGOAWAY") << 0 << int(Http2::INTERNAL_ERROR) << QNetworkReply::InternalServerError;
+    QTest::newRow("DelayedGOAWAY") << 1000 << int(Http2::INTERNAL_ERROR) << QNetworkReply::InternalServerError;
+    QTest::newRow("Graceful") << 0 << int(Http2::HTTP2_NO_ERROR) << QNetworkReply::RemoteHostClosedError;
+    QTest::newRow("Unknown") << 0 << 500000 << QNetworkReply::ProtocolFailure;
 }
 
 void tst_Http2::goaway()
@@ -613,6 +617,8 @@ void tst_Http2::goaway()
     using namespace Http2;
 
     QFETCH(const int, responseTimeoutMS);
+    QFETCH(const int, goawayCode);
+    QFETCH(const QNetworkReply::NetworkError, expectedError);
 
     clearHTTP2State();
 
@@ -620,7 +626,7 @@ void tst_Http2::goaway()
     nRequests = 3;
 
     ServerPtr srv(newServer(defaultServerSettings, defaultConnectionType()));
-    srv->emulateGOAWAY(responseTimeoutMS);
+    srv->emulateGOAWAY(goawayCode, responseTimeoutMS);
     QMetaObject::invokeMethod(srv.data(), "startServer", Qt::QueuedConnection);
     runEventLoop();
 
@@ -647,7 +653,7 @@ void tst_Http2::goaway()
     // No request processed, no 'replyFinished' slot calls:
     QCOMPARE(nRequests, 0);
     for (const auto &reply : replies)
-        QCOMPARE(reply->error(), QNetworkReply::InternalServerError);
+        QCOMPARE(reply->error(), expectedError);
     // Our server did not bother to send anything except a single GOAWAY frame:
     QVERIFY(!prefaceOK);
     QVERIFY(!serverGotSettingsACK);
-- 
2.51.2

